runtime/memory.c

```diff
@@ -15,648 +15,347 @@
-    if (caml_page_table_resize() != 0) return -1;
-  }
-  h = Hash(Page(page));
-  while (1) {
-    if (caml_page_table.entries[h] == 0) {
-      caml_page_table.entries[h] = page | toset;
-      caml_page_table.occupancy++;
-      break;
-    }
-    if (Page_entry_matches(caml_page_table.entries[h], page)) {
-      caml_page_table.entries[h] =
-        (caml_page_table.entries[h] & ~toclear) | toset;
-      break;
-    }
-    h = (h + 1) & caml_page_table.mask;
-  }
-  return 0;
-}
-
-#else
-
-/* 32-bit implementation:
-   The page table is represented as a 2-level array of unsigned char */
-
-CAMLexport unsigned char * caml_page_table[Pagetable1_size];
-static unsigned char caml_page_table_empty[Pagetable2_size] = { 0, };
-
-int caml_page_table_initialize(mlsize_t bytesize)
-{
-  int i;
-  for (i = 0; i < Pagetable1_size; i++)
-    caml_page_table[i] = caml_page_table_empty;
-  return 0;
-}
-
-static int caml_page_table_modify(uintnat page, int toclear, int toset)
+Caml_inline void write_barrier(
+  value obj, intnat field, value old_val, value new_val)
 {
-  uintnat i = Pagetable_index1(page);
-  uintnat j = Pagetable_index2(page);
-
-  if (caml_page_table[i] == caml_page_table_empty) {
-    unsigned char * new_tbl = caml_stat_calloc_noexc(Pagetable2_size, 1);
-    if (new_tbl == 0) return -1;
-    caml_page_table[i] = new_tbl;
-  }
-  caml_page_table[i][j] = (caml_page_table[i][j] & ~toclear) | toset;
-  return 0;
-}
+  /* HACK: can't assert when get old C-api style pointers
+    CAMLassert (Is_block(obj)); */
 
-#endif
-
-int caml_page_table_add(int kind, void * start, void * end)
-{
-  uintnat pstart = (uintnat) start & Page_mask;
-  uintnat pend = ((uintnat) end - 1) & Page_mask;
-  uintnat p;
+  if (!Is_young(obj)) {
 
-  for (p = pstart; p <= pend; p += Page_size)
-    if (caml_page_table_modify(p, 0, kind) != 0) return -1;
-  return 0;
+    if (Is_block(old_val)) {
+       /* if old is in the minor heap,
+          then this is in a remembered set already */
+       if (Is_young(old_val)) return;
+       /* old is a block and in the major heap */
+       caml_darken(0, old_val, 0);
+     }
+     /* this update is creating a new link from major to minor, remember it */
+     if (Is_block_and_young(new_val)) {
+       Ref_table_add(&Caml_state->minor_tables->major_ref, Op_val(obj) + field);
+     }
+   }
 }
 
-int caml_page_table_remove(int kind, void * start, void * end)
+CAMLexport CAMLweakdef void caml_modify (value *fp, value val)
 {
-  uintnat pstart = (uintnat) start & Page_mask;
-  uintnat pend = ((uintnat) end - 1) & Page_mask;
-  uintnat p;
+  write_barrier((value)fp, 0, *fp, val);
 
-  for (p = pstart; p <= pend; p += Page_size)
-    if (caml_page_table_modify(p, kind, 0) != 0) return -1;
-  return 0;
+  /* See Note [MM] above */
+  atomic_thread_fence(memory_order_acquire);
+  atomic_store_explicit(&Op_atomic_val((value)fp)[0], val,
+                        memory_order_release);
```

Comment:
Following my comment above based on the PLDI paper here I expect `memory_order_relaxed` instead of `memory_order_release`